### PR TITLE
fix: KT-45522 Define an object expression vs. anonymous type

### PR DESCRIPTION
**Pull request by p7nov**
Doc improvement per feedback in https://youtrack.jetbrains.com/issue/KT-45522

### CODE REVIEW

**Review by PetukhovVictor — APPROVED** [review]

File: docs/topics/object-declarations.md [hunk 1]
```diff
@@ -5,7 +5,37 @@ Kotlin handles this case with *object expressions* and *object declarations*.
 
 ## Object expressions
 
-To create an object of an anonymous class that inherits from some type (or types), write:
+_Object expressions_ create objects of anonymous classes, that is, classes that aren't explicitly declared with the `class`
+declaration. Such classes are handy for one-time use. You can inherit them from existing classes or interfaces or define
```
Comment (koshachy) [inline docs/topics/object-declarations.md:9]: Is it OK to say "inherit" for interfaces when we actually implement them?

File: docs/topics/object-declarations.md [hunk 2]
```diff
@@ -5,7 +5,37 @@ Kotlin handles this case with *object expressions* and *object declarations*.
...
+//sampleEnd
+    print(helloWorld)
+}
+```
+{kotlin-runnable="true"}
+
+### Inheriting anonymous objects from supertypes
+
+To create an object of an anonymous class that inherits from some type (or types), specify this type after `object` and
+colon (`:`). Then implement or override the members of this class as if you were [inheriting](inheritance.md) from it.
```
Comment (koshachy) [inline docs/topics/object-declarations.md:38]:
```suggestion
colon (`:`). Then implement or override the members of this class as if you were [inheriting](inheritance.md) from it:
```

File: docs/topics/object-declarations.md [hunk 3]
```diff
@@ -30,42 +60,59 @@ val ab: A = object : A(1), B {
...
-return type of a public function or the type of a public property, the actual type of that function or property
-will be the declared supertype of the anonymous object, or `Any` if you haven't declared any supertype. Members added
-in the anonymous object will not be accessible.
+If this function or property is public or private inline, its actual type is:
+* `Any` if the anonymous object doesn't have a declared supertype
+* the declared supertype of the anonymous object if there is exactly one such type 
+* the explicitly declared type if there is more than one declared supertype
+
+In all these cases, members added in the anonymous object are not accessible. Overriden members are accessible if they 
+are declared in the actual type of the function or property.
```
Comment (koshachy) [inline docs/topics/object-declarations.md:86]: As far as I could understand, the following example extends this sentence, so maybe we should put the colon.

File: docs/topics/object-declarations.md [hunk 4]
```diff
@@ -30,42 +60,59 @@ val ab: A = object : A(1), B {
...
+    fun getObjectB(): B = object: A, B { // explicit return type is required
+        override fun funFromA() {}
+        val x: String = "x"
     }
 }
 ```
 
+### Accessing variables from anonymous objects 
+
 The code in object expressions can access variables from the enclosing scope.
```
Comment (koshachy) [inline docs/topics/object-declarations.md:116]:
```suggestion
The code in object expressions can access variables from the enclosing scope:
```

**Review by koshachy — APPROVED** [review]

**Pull request merged**